pkg/detectors/azure_openai/azure_openai.go

```diff
@@ -0,0 +1,161 @@
+	dataStr := string(data)
+
+	// De-duplicate results.
+	tokens := make(map[string]struct{})
+	for _, match := range azureKeyPat.FindAllStringSubmatch(dataStr, -1) {
+		tokens[match[1]] = struct{}{}
+	}
+	if len(tokens) == 0 {
+		return
+	}
+	urls := make(map[string]struct{})
+	for _, match := range azureUrlPat.FindAllStringSubmatch(dataStr, -1) {
+		urls[match[1]] = struct{}{}
+	}
+
+	// Process results.
+	logCtx := logContext.AddLogger(ctx)
+	for token := range tokens {
+		s1 := detectors.Result{
+			DetectorType: s.Type(),
+			Redacted:     token[:3] + "..." + token[25:],
+			Raw:          []byte(token),
+		}
+
+		for url := range urls {
+			if verify {
+				client := s.client
+				if client == nil {
+					client = common.SaneHttpClient()
+				}
+
+				isVerified, extraData, verificationErr := verifyAzureToken(logCtx, client, url, token)
+				if isVerified || len(urls) == 1 {
+					s1.Verified = isVerified
+					s1.ExtraData = extraData
+					s1.SetVerificationError(verificationErr, token)
+					break
+				}
+			}
+		}
+
+		if s1.Verified {
+			results = append(results, s1)
+		} else {
+			if detectors.IsKnownFalsePositive(string(s1.Raw), detectors.DefaultFalsePositives, true) {
+				continue
+			}
+			results = append(results, s1)
+		}
+	}
+	return
+}
+
+func verifyAzureToken(ctx logContext.Context, client *http.Client, baseUrl, token string) (bool, map[string]string, error) {
+	// TODO: Replace this with a more suitable long-term endpoint.
+	// Most endpoints require additional info, e.g., deployment name, which complicates verification.
+	// This may be retired in the future, so we should look for another candidate.
+	// https://learn.microsoft.com/en-us/answers/questions/1371786/get-azure-openai-deployments-in-api
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("https://%s/openai/deployments?api-version=2023-03-15-preview", baseUrl), nil)
+	if err != nil {
+		return false, nil, nil
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("api-key", token)
+	res, err := client.Do(req)
+	if err != nil {
+		return false, nil, err
+	}
+	defer func() {
+		_, _ = io.Copy(io.Discard, res.Body)
+		_ = res.Body.Close()
+	}()
+
+	switch res.StatusCode {
+	case http.StatusOK:
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			return false, nil, err
+		}
+
+		var deployments deploymentsResponse
+		if err := json.Unmarshal(body, &deployments); err != nil {
+			if json.Valid(body) {
+				return false, nil, fmt.Errorf("failed to decode response %s: %w", req.URL, err)
+			} else {
+				// If the response isn't JSON it's highly unlikely to be valid.
+				return false, nil, nil
+			}
+		}
+
+		// JSON unmarshal doesn't check whether the structure actually matches.
+		if deployments.Object == "" {
+			return false, nil, nil
+		}
+
+		extraData := map[string]string{
+			"base_url": "https://" + baseUrl,
```

Comment:
Should the URL be extradata or `rawv2`?

Reply:
We use rawV2 to deduplicate findings, so rawV2 should have the key+url. Extra data is for including any extra helpful information.

Reply:
Gotcha. I've updated this to pass it in RawV2.